fem/libceed/ceed-assemble.cpp

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "ceed-assemble.hpp"
+
+#ifdef MFEM_USE_CEED
+#include "ceedsolvers-utility.h"
+
+int CeedHackReallocArray(size_t n, size_t unit, void *p)
+{
+   *(void **)p = realloc(*(void **)p, n*unit);
+   if (n && unit && !*(void **)p)
+      return CeedError(NULL, 1, "realloc failed to allocate %zd members of size "
+                       "%zd\n", n, unit);
+   return 0;
+}
+
+#define CeedHackRealloc(n, p) CeedHackReallocArray((n), sizeof(**(p)), p)
+
+int CeedHackFree(void *p)
+{
+   free(*(void **)p);
+   *(void **)p = NULL;
+   return 0;
+}
```

Comment:
I want to use `CeedRealloc`, but it is marked with hidden visibility. I am very open to other solutions here, I do not very much like copy/pasting this code from Ceed here. See also (https://github.com/mfem/mfem/pull/1763#discussion_r522518340)

Reply:
Why do you want `CeedRealloc`? We made it internal because we didn't want libCEED memory management idioms to leak out (or to encourage callers to adopt them). I'd think either MFEM has/can have a realloc idiom or it wouldn't be needed in this interface. Of course it's a different matter if the intent is to develop and test this code here and then submit to libCEED.

Reply:
The code that uses `Realloc` is precisely the code (explicit assembly of a `CeedOperator`, used on the coarsest level so we can use conventional AMG) that we want to eventually see in libCEED itself. One option is to use a more MFEM-y idiom here and then change it again when submitting to libCEED, another is to try to push it in libCEED before we merge this in MFEM, but I am lazy and the code works so I am inclined to leave it more or less as is.

Reply:
In that case, I think it's fine to do what you've done since this code won't live in MFEM for long.